Krypton/LocalNotificationAuthority.swift

```diff
@@ -0,0 +1,130 @@
+//
+//  LocalNotificationAuthority.swift
+//  Krypton
+//
+//  Created by Alex Grinman on 1/7/18.
+//  Copyright © 2018 KryptCo. All rights reserved.
+//
+
+import Foundation
+import Sodium
+import JSON
+
+class LocalNotificationAuthority {
+    
+    internal struct SignedLocalRequest {
+        let localRequestJSON:String
+        let signature:Data
+    }
+    
+    enum Errors:Error {
+        case invalidLocalNotificationSignature
+        case signingKeyGenFailed
+        case signatureCreationFailed
+        case badLocalRequestSignature
+    }
+    
+    class VerifiedLocalRequest:LocalRequestBody {}
+    class UnverifiedLocalRequest:LocalRequestBody {}
+    
+    private static let signingKeyIdentifier = "local_notification_signing_key"
+    
+    static func createSigningKeyIfNeeded() throws {
+        do {
+            let _ = try getSigningKey()
+        } catch KeychainStorageError.notFound {
+            // create the key
+            log("local notification signing key not found...creating one now", .warning)
+            
+            guard let signingKey = KRSodium.instance().auth.key() else {
+                throw Errors.signingKeyGenFailed
+            }
+            
+            try KeychainStorage().setData(key: signingKeyIdentifier, data: signingKey)
+        }
+    }
+    
+    // sodium signing key (Auth SecretKey)
+    private static func getSigningKey() throws -> Auth.SecretKey {
+        let signingKey = try KeychainStorage().getData(key: signingKeyIdentifier)
+        return signingKey
+    }
+    
+    static func createSignedPayload(for verifiedLocalRequest:VerifiedLocalRequest) throws -> [String:Any] {
+        
+        let jsonString = try verifiedLocalRequest.jsonString()
+        let jsonData = try jsonString.utf8Data()
+        let signingKey = try getSigningKey()
+        
+        guard let signature = KRSodium.instance().auth.tag(message: jsonData, secretKey: signingKey) else {
+            throw Errors.signatureCreationFailed
+        }
+        
+        let localRequest = SignedLocalRequest(localRequestJSON: jsonString, signature: signature)
+        return localRequest.object
+    }
+    
+    static func verifiedLocalNotification(with payload:[String:Any]) throws -> VerifiedLocalRequest {
+        
+        let rawRequest = try SignedLocalRequest(json: payload)
+        let rawRequestData = try rawRequest.localRequestJSON.utf8Data()
+        
+        let signingKey = try getSigningKey()
+        
+        guard KRSodium.instance().auth.verify(message: rawRequestData, secretKey: signingKey, tag: rawRequest.signature)
+        else {
+            throw Errors.badLocalRequestSignature
+        }
+        
+        let verifiedLocalRequest = try VerifiedLocalRequest(jsonData: rawRequestData)
+        return verifiedLocalRequest
+    }
+    
+    static func unverifiedLocalNotification(with payload:[String:Any]) throws -> UnverifiedLocalRequest {
```

Comment:
I think `parseUnverifiedLocalNotification` is more clear of a name

Reply:
👍